hypothesis-python/src/hypothesis/strategies/_internal/strings.py

```diff
@@ -78,40 +73,7 @@ def __repr__(self):
         return self._force_repr or f"OneCharStringStrategy({self.intervals!r})"
 
     def do_draw(self, data):
-        if len(self.intervals) > 256:
-            if biased_coin(data, 0.2):
-                i = integer_range(data, 256, len(self.intervals) - 1)
-            else:
-                i = integer_range(data, 0, 255)
-        else:
-            i = integer_range(data, 0, len(self.intervals) - 1)
-
-        i = self.rewrite_integer(i)
-
-        return chr(self.intervals[i])
-
-    def rewrite_integer(self, i):
-        # We would like it so that, where possible, shrinking replaces
-        # characters with simple ascii characters, so we rejig this
-        # bit so that the smallest values are 0, 1, 2, ..., Z.
-        #
-        # Imagine that numbers are laid out as abc0yyyZ...
-        # this rearranges them so that they are laid out as
-        # 0yyyZcba..., which gives a better shrinking order.
-        if i <= self.Z_point:
-            # We want to rewrite the integers [0, n] inclusive
-            # to [zero_point, Z_point].
-            n = self.Z_point - self.zero_point
-            if i <= n:
-                i += self.zero_point
-            else:
-                # We want to rewrite the integers [n + 1, Z_point] to
-                # [zero_point, 0] (reversing the order so that codepoints below
-                # zero_point shrink upwards).
-                i = self.zero_point - (i - n)
-                assert i < self.zero_point
-            assert 0 <= i <= self.Z_point
-        return i
+        return data.draw_string(self.intervals, min_size=1, max_size=1)
```

Comment:
I'm uncertain about this. `OneCharStringStrategy` always drawing a single string effectively makes our IR node `char`, not `string`...

Indeed, in this pull, currently nothing uses `PrimitiveProvider.draw_string` except for drawing single characters. Conceptually, it would be nice if `TextStrategy` called `PrimitiveProvider.draw_string`. However, as consumers can pass any arbitrary `alphabet: SearchStrategy[str]` to `text()`, we can't rely on the precomputed existence of intervals (or, in fact, assume anything about the distribution of the strategy). So delegating to `PrimitiveProvider.draw_string` and ignoring the provided strategy would change semantics.

Would be grateful for guidance here - I'm not sure what the best way forward is.

Reply:
I think we can redefine `TextStrategy.do_draw()` as

```python
class TextStrategy(ListStrategy):
    def do_draw(self, data):
        if isinstance(elems := unwrap_strategies(self.element_strategy), OneCharStringStrategy):
            return data.draw_string(elems.intervals, min_size=self.min_size, max_size=self.max_size)
        return "".join(super().do_draw(data))
```
so that the former case gives us a nice IR, and the latter handles annoying custom alphabet strategies which will consist of more granular IR pieces.

We could even go further, and make `OneCharStringStrategy` a subclass of `TextStrategy` which hardcodes that min_size = max_size = 1 and intervals is not None, but I think that's better left for a future PR.